clang/compiler-course/dudchenko_o_ast_var_1/dudchenko_o_ast_var_1.cpp

```diff
@@ -0,0 +1,123 @@
+#include "clang/AST/ASTConsumer.h"
+#include "clang/AST/RecursiveASTVisitor.h"
+#include "clang/Frontend/CompilerInstance.h"
+#include "clang/Frontend/FrontendPluginRegistry.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+class TypeInfoVisitor : public clang::RecursiveASTVisitor<TypeInfoVisitor> {
+public:
+  explicit TypeInfoVisitor(clang::ASTContext *context) {}
+
+  bool VisitCXXRecordDecl(clang::CXXRecordDecl *record) {
+    if (record->isImplicit()) {
+      return true;
+    }
+
+    llvm::outs() << record->getName();
+
+    if (record->getNumBases() > 0) {
+      llvm::outs()
+          << " -> "
+          << record->bases_begin()->getType()->getAsCXXRecordDecl()->getName();
+    }
+
+    llvm::outs() << " \n";
+
+    if (record->field_begin() != record->field_end()) {
+      llvm::outs() << "|_Fields\n";
+      for (const auto *field : record->fields()) {
+        llvm::outs() << "| |_ " << field->getName() << " ("
+                     << field->getType().getAsString() << "|"
+                     << getAccessSpecifierString(field->getAccess()) << ")\n";
+      }
+    }
+
+    llvm::outs() << "|\n";
+
+    if (record->method_begin() != record->method_end()) {
+      llvm::outs() << "|_Methods\n";
+      bool firstMethod = true;
+      for (const auto *method : record->methods()) {
+        if (method->isImplicit() || method->isDefaulted()) {
+          continue;
+        }
+
+        if (firstMethod) {
+          firstMethod = false;
+        }
+
+        llvm::outs() << "| |_ " << method->getNameAsString() << " ("
+                     << method->getReturnType().getAsString() << "()|"
+                     << getAccessSpecifierString(method->getAccess());
+
+        bool isVirtualMethod = method->isVirtual();
+        bool isPureVirtual = method->isPureVirtual();
+        bool hasOverride = method->size_overridden_methods() > 0;
+
+        if (isVirtualMethod) {
+          if (hasOverride) {
+            llvm::outs() << "|override";
+          } else {
+            llvm::outs() << "|virtual";
+            if (isPureVirtual) {
+              llvm::outs() << "|pure";
+            }
+          }
+        } else {
+          if (isPureVirtual) {
+            llvm::outs() << "|pure";
+          }
```

Comment:
I guess, if the method is not virtual it cannot be pure & virtual at the same time either